py/dml/c_backend.py

```diff
@@ -2623,6 +2814,173 @@ def register_saved_attributes(initcode, node):
                  postindent = -1)
     initcode.out('}\n', preindent=-1)
     initcode.out('}\n', preindent=-1)
+
+def generate_hook_attr_names(node, prefix):
+    # Generate attributes for all subcomponents.
+    for child in sorted(node.get_components(),
+                        key=lambda o: o.name or ''):
+        if child.objtype == 'hook':
+            yield (child, prefix + child.ident)
+        elif (isinstance(child, objects.CompositeObject)
+              and child.objtype not in {'bank', 'port', 'subdevice'}):
+            yield from generate_hook_attr_names(
+                child, prefix + param_str(child, 'name') + "_")
+
+def register_hook_attributes(initcode, node):
+    if node is dml.globals.device:
+        for port in node.get_components('bank', 'port', 'subdevice'):
+            register_hook_attributes(initcode, port)
+        cls = 'class'
+        getter = '_DML_get_hook_attr'
+        setter = '_DML_set_hook_attr'
+    else:
+        if dml.globals.dml_version == (1, 2) and not node.name:
+            # anonymous bank
+            return
```

Comment:
ICE?

Reply:
oh, never mind, it's done to stop recursion.

There's a anonymous bank corner that deserves ICE,

```
dml 1.2;
import "14.dml";
bank { is foo; }
```
14.dml:
```
dml 1.4;
template foo {
  hook int x;
}
 ```
but I guess it's so obscure that we can accept it's silently broken.

However, anonymous fields are a bigger problem, maybe you can write something like
```
dml 1.2;
import "14.dml";
bank b { register r size 4 @ 0; }
```
14.dml:
```
dml 1.4;
in each field { hook int x; }
```
where I suppose the hooks will be silently broken using your patch.

So I suppose we should write `# anonymous bank or field` here, and perhaps add a 1.2-specific explicit ICE or error to reject hooks in anonymous objects.

Reply:
You're quite right, it does indeed get broken. Time to figure out the best way to catch that and ICE about it then...

Reply:
We should do the same re. saved variables.

Reply:
The issue re. fields is not connected to this line, though -- in this context the node we're studying has to correspond to a Simics object, i.e. it's a `device`, `subdevice`, `bank`, or `port` and of those only banks may be anonymous.

Reply:
ah right. This can be expressed as a trusted comment `assert isinstance(obj, Bank)`

Reply:
Have now sprinkled some ICEing where relevant.